Run commands asynchronously

komuta opened the pull request:
Performance wise, it is critical to avoid blocking some callbacks (typically, `Isolator::usage()`) while running the associated command. Mesos is designed to allow those callbacks to run concurrently, asynchronously. This PR leverages libprocess' primitive to avoid typically pitfall of synchronous calls.

komuta committed:
Add CommandRunner::asyncRun

And make CommandRunner::run use it synchronously.

komuta committed:
Remove temporary debug output

komuta committed:
Fix tests execution

komuta committed:
Use asynchronous command for usage()

kamaradclimber commented on src/CommandRunner.cpp:
```diff
@@ -72,10 +78,10 @@ class TemporaryFile {
-    if (remove(m_filepath.c_str()) != 0)
-      std::cerr << "Error while deleting " << m_filepath << std::endl;
-  }
+  /* ~TemporaryFile() { */
```
commented code

kamaradclimber reviewed (approved):
overall looks ok but I think we should use it in the wild for a few hours to make sure it behaves as expected

Mwea commented on CMakeLists.txt:
```diff
@@ -119,5 +119,5 @@ find_path(GTEST_INCLUDE_DIR NAMES "gtest/gtest.h" PATHS ${GTEST_SRCH_PATH} NO_DE
 
 target_include_directories(${TEST_BINARY_NAME} PUBLIC ${GTEST_INCLUDE_DIR} tests)
 target_link_libraries(${TEST_BINARY_NAME} ${PROJECT_NAME} ${MESOS_BUILD_DIR}/3rdparty/.libs/libgmock.a pthread)
-add_custom_target(check COMMAND ${TEST_BINARY_NAME} ${CMAKE_SOURCE_DIR}/tests/scripts/)
```
Why did you need to do that ?

Mwea commented on CMakeLists.txt:
```diff
@@ -119,5 +119,5 @@ find_path(GTEST_INCLUDE_DIR NAMES "gtest/gtest.h" PATHS ${GTEST_SRCH_PATH} NO_DE
 
 target_include_directories(${TEST_BINARY_NAME} PUBLIC ${GTEST_INCLUDE_DIR} tests)
 target_link_libraries(${TEST_BINARY_NAME} ${PROJECT_NAME} ${MESOS_BUILD_DIR}/3rdparty/.libs/libgmock.a pthread)
-add_custom_target(check COMMAND ${TEST_BINARY_NAME} ${CMAKE_SOURCE_DIR}/tests/scripts/)
```
Why did you need to do that ?

komuta replied:
the test executable was using the first argument (optional) to specify the root directory for test scripts. But this make it incompatible with googletest command line interface. I changed that, so now we can override the default test script directory through an environment variable.

Mwea commented on scripts/travis.sh:
```diff
@@ -21,4 +21,4 @@ if [ -n "$TRAVIS" ]; then
 
 make
-make check
+TEST_RESOURCES_PATH=../tests/scripts/ make check
```
Same question as above ?

komuta replied:
Same answer ;)

Mwea commented on src/CommandIsolator.cpp:
```diff
@@ -48,6 +49,12 @@ class CommandIsolatorProcess : public process::Process<CommandIsolatorProcess> {
  private:
+  inline static ::mesos::ResourceStatistics emptyStats() {
+    ::mesos::ResourceStatistics stats;
+    stats.set_timestamp(Clock::now().secs());
```
I'd prefer to see timestamp passed as a parameter and you get clock::now().secs in the method usage

Mwea commented on src/CommandIsolator.cpp:
```diff
@@ -134,44 +141,30 @@ process::Future<ContainerLimitation> CommandIsolatorProcess::watch(
+      .asyncRun(m_usageCommand.get(), stringify(inputsJson))
+      .then([](Try<string> output) -> Future<::mesos::ResourceStatistics> {
+        if (output.isError() || output->empty()) {
+          LOG(WARNING) << "Unable to parse output: " << output.error();
```
Wont work if output is not an error, you should have dedicated messages

komuta replied:
You're right, it was actually crashing, and this has been fixed

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -72,10 +78,10 @@ class TemporaryFile {
-      std::cerr << "Error while deleting " << m_filepath << std::endl;
-  }
+  /* ~TemporaryFile() { */
+  /*   if (remove(m_filepath.c_str()) != 0) */
```
Do you really want to get rid of this ?

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -196,124 +202,85 @@ class Timer {
+    unsigned long timeoutInSeconds, const logging::Metadata& loggingMetadata) {
+  // FIXME: remove this
+  vector<string> full_args;
+  full_args.push_back(theCommand);
```
isn't there a method to push everything at once ?

komuta replied:
Sadly, no. It's C++, the standard lib is not really designed to make your code more digest...

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -196,124 +202,85 @@ class Timer {
+Future<Try<bool>> runCommandWithTimeout(
+    const std::string& theCommand, const std::vector<std::string>& args,
+    unsigned long timeoutInSeconds, const logging::Metadata& loggingMetadata) {
+  // FIXME: remove this
```
?

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -196,124 +202,85 @@ class Timer {
+  full_args.push_back(args[1]);
+  full_args.push_back(args[2]);
+  Try<Subprocess> command =
+      subprocess(theCommand, full_args, Subprocess::PATH(args[0]),
```
Maybe you could suggest a better name than theCommand ?

Mwea replied:
It doesn't say what is the difference between command and theCommand

komuta replied:
done

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -196,124 +202,85 @@ class Timer {
+  if (command.isError()) {
+    TASK_LOG(ERROR, loggingMetadata) << "Error launching external command \""
+                                     << theCommand << "\": " << command.error();
+    return Error("launch error");
```
Make a string variable containing the message , pass it to task log AND Error

komuta replied:
done

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -196,124 +202,85 @@ class Timer {
+  return process.status()
+      .then([=](Option<int> status) -> Future<Try<bool>> {
+        if (status.isNone()) {
+          TASK_LOG(ERROR, loggingMetadata)
```
Same suggestion as above

komuta replied:
done

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -196,124 +202,85 @@ class Timer {
+              << "Error getting status for external command \"" << theCommand
+              << "\"";
+          return Error("error getting command status");
+        } else if (status.get() != 0) {
```
if (rc > 0 && (WIFEXITED(status) || WIFSIGNALED(status)))  Why did you get rid of this ?

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -196,124 +202,85 @@ class Timer {
+        } else if (status.get() != 0) {
+          if (WIFSIGNALED(status.get()) && WTERMSIG(status.get()) != 0) {
+            int signalCode = WTERMSIG(status.get());
+            TASK_LOG(ERROR, loggingMetadata)
```
It feels like this whole stuff could have been migrated to dedicated methods so you re-use previous code

Mwea replied:
I'm not confident about that code, it feels like you might break the implementation

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -355,5 +322,18 @@ Try<string> CommandRunner::run(const Command& command,
+                               const std::string& input) {
+  Future<Try<string>> output = asyncRun(command, input);
+  auto result = await(output);
+  if (!result.await(Seconds(command.timeout() + 1))) {
```
Why ?

komuta replied:
I wanted to ensure that in the typical timeout case, the `after()` in `asyncRun()` is triggered before this `await()`expires. But I changed my copy, and now rely only to `asyncRun()` for timeout.

komuta committed:
Don't capture stdout and stderr

komuta committed:
Better with the test script...

komuta committed:
Small fixes

- rename theCommand in executable
- return proper error message
- don't expire `await()` in `run()` (rely on `asyncRun()` for it)

komuta committed:
Remove obsolete code

komuta committed:
Add error detail in the logs in case of failure

komuta committed:
Fix removal of temporary file after run

This will work whatever the outcome. Due to the async nature of the
call, we can't use the "remove file on destructor" trick anymore.

komuta committed:
Ensures temp files are deleted even on failure

komuta committed:
Fix force kill still running command

komuta committed:
Fix usage() crash on empty output

komuta committed:
Return future failure on command timeout

komuta committed:
Handle timeouts on usage()

komuta committed:
Terminate/kill the whole process tree

komuta committed:
Cosmetic

komuta committed:
Add missing test script *again*

komuta committed:
Remove temporary code

Mwea commented on src/CommandRunner.cpp:
```diff
@@ -72,248 +81,108 @@ class TemporaryFile {
  * TODO(clems4ever): split this method so that it becomes easier to read.
  *
- * @param command The command to execute in the child process.
+ * @param executable Absolute path to the executed of rhe command to execute in
```
typo

komuta committed:
Typo

Mwea commented on src/ConfigurationParser.cpp:
```diff
@@ -1,6 +1,7 @@
 #include "ConfigurationParser.hpp"
 
 #include <stout/foreach.hpp>
+#include <map>
```
Why ?

komuta replied:
just for linting, I could have skipped that

komuta committed:
Make emptyStats timestamp more consistent

Mwea commented:
LGTM
